cosmos.go

```diff
@@ -203,39 +217,133 @@ func (c *cosmosImpl) dial() (interfaces.QueryExecutor, error) {
 
 func (c *cosmosImpl) ExecuteQuery(query interfaces.QueryBuilder) ([]interfaces.Response, error) {
 	if query == nil {
-		return nil, fmt.Errorf("Query is nil")
+		return nil, fmt.Errorf("query is nil")
 	}
 	return c.Execute(query.String())
 }
 
 func (c *cosmosImpl) Execute(query string) ([]interfaces.Response, error) {
 
-	responses, err := c.pool.Execute(query)
+	doRetry := func() ([]interfaces.Response, error) {
+		return c.pool.Execute(query)
+	}
+
+	responses, err := c.retryLoop(doRetry)
 
 	// try to investigate the responses and to find out if we can find more specific error information
 	if respErr := extractFirstError(responses); respErr != nil {
 		err = respErr
 	}
 
-	updateRequestMetrics(responses, c.metrics)
 	return responses, err
 }
 
 func (c *cosmosImpl) ExecuteWithBindings(query string, bindings, rebindings map[string]interface{}) ([]interfaces.Response, error) {
 
-	responses, err := c.pool.ExecuteWithBindings(query, bindings, rebindings)
+	doRetry := func() ([]interfaces.Response, error) {
+		return c.pool.ExecuteWithBindings(query, bindings, rebindings)
+	}
+
+	responses, err := c.retryLoop(doRetry)
 
 	// try to investigate the responses and to find out if we can find more specific error information
 	if respErr := extractFirstError(responses); respErr != nil {
 		err = respErr
 	}
 
-	updateRequestMetrics(responses, c.metrics)
+	return responses, err
+}
+
+type retryFun func() ([]interfaces.Response, error)
+
+func (c *cosmosImpl) retryLoop(executeRetry retryFun) (responses []interfaces.Response, err error) {
+	var retryCount int
+
+	for retryCount = 0; retryCount < c.maxRetries+1; retryCount++ {
```

Comment:
I would suggest to introduce a variable

`maxTries := c.maxRetries+1`
and then
`for tryCount = 0; tryCount < maxTries; tryCount++ {`

is easier to understand